### PR TITLE
Victor/eco 805/jwt validation

### PR DESCRIPTION
**Pull request by Doc999tor**

### CODE REVIEW

File: log [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (doodyparizada) [inline log]: committed by accident?

File: package-lock.json [hunk 2]
```diff
@@ -1205,7 +1205,7 @@
 			"dependencies": {
 				"assign-symbols": {
 					"version": "0.1.1",
-					"resolved": "https://registry.npmjs.org/assign-symbols/-/assign-symbols-0.1.1.tgz",
+					"resolved": "http://registry.npmjs.org/assign-symbols/-/assign-symbols-0.1.1.tgz",
```
Comment (doodyparizada) [inline package-lock.json]: why are you replacing https with http?

File: scripts/src/errors.ts [hunk 3]
```diff
@@ -46,7 +46,8 @@ const CODES = {
 		MaxWalletsExceeded: 7,
 		InvalidWalletAddress: 8,
 		ExpiredJwt: 9,
-		InvalidJwtIssuedTime: 10
+		InvalidJwtIssuedTime: 10,
+		MissingFieldJWT: 11
```
Comment (doodyparizada) [inline scripts/src/errors.ts]: Add to error-cods wiki

File: scripts/src/public/services/native_offers.ts [hunk 4]
```diff
@@ -44,19 +44,42 @@ export function isPayToUser(jwt: ExternalOrderJWT): jwt is ExternalPayToUserOrde
...
 	if (decoded.payload.sub !== "earn" && decoded.payload.sub !== "spend" && decoded.payload.sub !== "pay_to_user") {
 		throw InvalidExternalOrderJwt();
 	}
 
-	if ((decoded.payload.sub === "spend" || decoded.payload.sub === "pay_to_user") &&
-		!!decoded.payload.sender.user_id && decoded.payload.sender.user_id !== appUserId) {
+	switch (decoded.payload.sub) {
+		case "spend":
+			if (!decoded.payload.sender) { throw MissingFieldJWT("sender"); }
+			if (!decoded.payload.offer) { throw MissingFieldJWT("offer"); }
```
Comment (doodyparizada) [inline scripts/src/public/services/native_offers.ts]: you can check `offer` out of the case

File: scripts/src/public/services/native_offers.ts [hunk 5]
```diff
@@ -44,19 +44,42 @@ export function isPayToUser(jwt: ExternalOrderJWT): jwt is ExternalPayToUserOrde
...
+		(decoded.payload.sub === "spend" || decoded.payload.sub === "pay_to_user")
+		&& decoded.payload.sender && !!decoded.payload.sender.user_id
+		&& decoded.payload.sender.user_id !== appUserId
+	) {
 		// if sender.user_id is defined and is different than current user, raise error
 		throw ExternalEarnOfferByDifferentUser(appUserId, decoded.payload.sender.user_id);
 	}
 
-	if (decoded.payload.sub === "earn" && decoded.payload.recipient.user_id !== appUserId) {
+	if (decoded.payload.sub === "earn" && decoded.payload.recipient && decoded.payload.recipient.user_id !== appUserId) {
```
Comment (doodyparizada) [inline scripts/src/public/services/native_offers.ts]: I think it should be:
```typescript
decoded.payload.sub === "earn" && (!decoded.payload.recipient || decoded.payload.recipient.user_id !== appUserId)
```

File: scripts/src/errors.ts (hunk 3 above)
Comment (nitzantomer) [reply scripts/src/errors.ts]: here: https://github.com/kinecosystem/marketplace-server/wiki/Error-Codes

File: scripts/src/public/services/native_offers.ts (hunk 5 above)
Comment (nitzantomer) [reply scripts/src/public/services/native_offers.ts]: @doodyparizada 
no, this is actually correct. he checks if `decoded.payload.recipient` is missing above (and if so throws a different error: `MissingFieldJWT("recipient");`

Comment (doodyparizada) [reply scripts/src/public/services/native_offers.ts]: So remove `&& decoded.payload.recipient ` check completely. It being here is wrong. If the check above is removed, this code will allow earn JWT that are missing the `recipient` key.

File: scripts/src/errors.ts (hunk 3 above)
Comment (Doc999tor) [reply scripts/src/errors.ts]: Added

File: package-lock.json (hunk 2 above)
Comment (Doc999tor) [reply package-lock.json]: Yes, by accident. It's automatic updates, I haven't touched it manually

File: scripts/src/public/services/native_offers.ts (hunk 5 above)
Comment (Doc999tor) [reply scripts/src/public/services/native_offers.ts]: Removed

File: scripts/src/public/services/native_offers.ts (hunk 4 above)
Comment (doodyparizada) [reply scripts/src/public/services/native_offers.ts]: Yes. We'll push it back in if that happens. For now let's reduce duplication.

File: scripts/src/public/services/native_offers.ts (hunk 5 above)
Comment (doodyparizada) [reply scripts/src/public/services/native_offers.ts]: 👍

File: package-lock.json (hunk 2 above)
Comment (doodyparizada) [reply package-lock.json]: Are you using npm 6.5.0?

Comment (Doc999tor) [reply package-lock.json]: No, 6.4.1. I'll upgrade it
--

**Pull request merged**